### PR TITLE
[SPARK-20664][core] Delete stale application data from SHS.

### PR DESCRIPTION
**Pull request by vanzin**
Detect the deletion of event log files from storage, and remove
data about the related application attempt in the SHS.

Also contains code to fix SPARK-21571 based on code by ericvandenbergfb.

### CODE REVIEW

**Comment by vanzin** [comment]
@ericvandenbergfb

**Comment by vanzin** [comment]
There are some previous comments on this code at: https://github.com/vanzin/spark/pull/40

**Comment by vanzin** [comment]
@squito

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 1]
```diff
@@ -806,6 +867,17 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
       throw new NoSuchElementException(s"Cannot find attempt $attemptId of $appId."))
   }
 
+  private def deleteLog(log: Path): Unit = {
+    try {
+      fs.delete(log, true)
+    } catch {
+      case e: AccessControlException =>
```
Comment (gengliangwang) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: Nit: `e` is not used

**Review by gengliangwang — COMMENTED** [review]
Overall LGTM since I have reviewed in https://github.com/vanzin/spark/pull/40

**Review by squito — COMMENTED** [review]
still looking, but on the reviews of @ericvandenbergfb 's changes, it seemed like @ajbozarth and @jiangxb1987 were opposed to the more aggressive cleaning by default.  I don't see the argument against it, but want to make sure they are aware of that change here.

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 2]
```diff
@@ -405,49 +404,70 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
...
-        for (file <- logInfos) {
-          tasks += replayExecutor.submit(new Runnable {
-            override def run(): Unit = mergeApplicationListing(file)
+      val tasks = updated.map { entry =>
+        try {
+          replayExecutor.submit(new Runnable {
+            override def run(): Unit = mergeApplicationListing(entry, newLastScanTime)
           })
+        } catch {
+          // let the iteration over logInfos break, since an exception on
```
[the remark this answers is not in the document]
Comment (squito) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: and actually you've moved the try/catch so this is no longer true, you'll continue to submit all tasks if one throws an exception.  (I guess I'm not really sure why the old code did it that way ...)

**Comment by ajbozarth** [comment]
I haven't had a chance to read though your code, but as @squito said, I am against any default feature that deletes files from the eventLog dir. Many users, such as myself, use one log dir for both the event log as well as their Spark logs. I believe it is a great feature for most use cases and should be available as a option defaulted to off.

**Comment by squito** [comment]
well, perhaps I mis-represented this -- you still need to turn the event log cleaning on explicitly with the old option, "spark.history.fs.cleaner.enabled".  This just doesn't include the "aggressive" option that was originally proposed by @ericvandenbergfb

**Comment by ajbozarth** [comment]
Ok, no problems here on that front then. If I have time later to do a proper review and this has't been merged yet I'll take better a look at the whole PR

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 3]
```diff
@@ -544,73 +621,75 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
...
+          }
         }
-      }
 
-      addListing(app)
+        addListing(app)
+        (Some(app.info.id), app.attempts.head.info.attemptId)
+
+      case _ =>
+        (None, None)
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: I think comment here explaining that writing an entry with no appId will mark this log file as eligible for automatic recovery, if its still in that state after max_log_age.  (if I understood correctly)

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 4]
```diff
@@ -834,6 +906,9 @@ private[history] case class FsHistoryProviderMetadata(
 
 private[history] case class LogInfo(
     @KVIndexParam logPath: String,
+    @KVIndexParam("lastProcessed") lastProcessed: Long,
+    appId: Option[String],
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: also a comment here explaining why appId is an Option, as that is unexpected

**Comment by jiangxb1987** [comment]
I was actually suggesting have the "aggressive" option default turned on, and I'm also fine to not have that config at all. Will take a closer look at this later, thanks you for ping me @squito !

File: core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala [hunk 5]
```diff
@@ -663,6 +665,95 @@ class FsHistoryProviderSuite extends SparkFunSuite with BeforeAndAfter with Matc
...
+    // Delete the second attempt's log file. Now everything should go away.
+    doReturn(3L).when(provider).getNewLastScanTime()
+    attempt2.delete()
+    updateAndCheck(provider) { list =>
+      assert(list.isEmpty)
+    }
+  }
+
+  test("SPARK-21571: clean up removes invalid history files") {
+    val clock = new ManualClock(TimeUnit.DAYS.toMillis(120))
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala]: just curious, why start at 120 days?  (not that it matters ...)

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (jiangxb1987) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: Maybe we should handle `RejectedExecutionException` explicitly, under this exception, we can log an error message and stop submit the rest tasks.

File: core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala [hunk 6]
```diff
@@ -663,6 +665,95 @@ class FsHistoryProviderSuite extends SparkFunSuite with BeforeAndAfter with Matc
...
+    provider.checkForLogs()
+    provider.cleanLogs()
+    assert(new File(testDir.toURI).listFiles().size === 3)
+
+    // Move the clock forward another 2 days and scan the files again. This time the cleaner should
+    // pick up the invalid files and get rid of them.
+    clock.advance(TimeUnit.DAYS.toMillis(2))
+    provider.checkForLogs()
+    provider.cleanLogs()
+    assert(new File(testDir.toURI).listFiles().size === 0)
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala]: I think you should add a case where one file starts out empty, say even for one full day, but then becomes valid before the expiration time, and make sure it does *not* get cleaned up.

**Review by squito — APPROVED** [review]
one small suggestion for an additional test, otherwise lgtm

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: Actually `RejectedExecutionException` shouldn't ever be thrown here. The executor doesn't have a bounded queue, and it's very unlikely you'll ever submit `Integer.MAX_VALUE` tasks here.

The code didn't use to catch any exception here (it was added along with the comment in a531fe1). Catching the exception doesn't do any harm, I just don't think this code will ever trigger.

File: core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala (hunk 5 above)
Comment (vanzin) [reply core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala]: This line:

    val maxTime = clock.getTimeMillis() - conf.get(MAX_LOG_AGE_S) * 1000

Without that `maxTime` would be negative and that seems to be triggering a bug somewhere else. I need to take a look at exactly what's happening there, but it seems unrelated to this change.

**Comment by squito** [comment]
lgtm

**Comment by jiangxb1987** [comment]
LGTM

**Comment by jiangxb1987** [comment]
This PR introduces an more thoughtful event logs cleanup method, if users have `spark.history.fs.cleaner.enabled` set to true, their existing invalid event log files(empty/no valid applicationId) that have past the max log age can now be deleted. It should be safe to do so because users should not have files other than event log files in the directory. Can we merge this to master/2.3? @sameeragarwal

**Comment by gengliangwang** [comment]
LGTM and it looks safe.

File: core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala (hunk 5 above)
Comment (vanzin) [reply core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala]: FYI #20284 fixes the underlying bug.

**Comment by squito** [comment]
as RC1 failed and RC2 is going to be cut soon, I'm going to merge this to master & 2.3

**Pull request closed without merging**